# MCIS6273 Data Mining (Prof. Maull) / Fall 2021 / HW0

**This assignment is worth up to 20 POINTS to your grade total if you complete it on time.**

| Points <br/>Possible | Due Date | Time Commitment <br/>(estimated) |
|:---------------:|:--------:|:---------------:|
| 20 | Wednesday, Sep 1 @ Midnight | _up to_ 20 hours |


* **GRADING:** Grading will be aligned with the completeness of the objectives.

* **INDEPENDENT WORK:** Copying, cheating, plagiarism  and academic dishonesty _are not tolerated_ by University or course policy.  Please see the syllabus for the full departmental and University statement on the academic code of honor.

## OBJECTIVES
* Familiarize yourself with the JupyterLab environment, Markdown and Python

* Familiarize yourself with Github and basic git

* Explore JupyterHub Linux console integrating what you learned in the prior parts of this homework

* Listen to the Talk Python['Podcast'] from June 25, 2021: A Path to Data Science Interview with Sanyam Bhutani

* Explore Python for data munging and analysis, with an introduction to CSV and Pandas

## WHAT TO TURN IN
You are being encouraged to turn the assignment in using the provided
Jupyter Notebook.  To do so, make a directory in your Lab environment called
`homework/hw0`.   Put all of your files in that directory.  Then zip that directory,
rename it with your name as the first part of the filename (e.g. `maull_hw0_files.zip`), then
download it to your local machine, then upload the `.zip` to Blackboard.

If you do not know how to do this, please ask, or visit one of the many tutorials out there
on the basics of using zip in Linux.

If you choose not to use the provided notebook, you will still need to turn in a
`.ipynb` Jupyter Notebook and corresponding files according to the instructions in
this homework.


## ASSIGNMENT TASKS
### (0%) Familiarize yourself with the JupyterLab environment, Markdown and Python 

As stated in the course announcement [Jupyter (https://jupyter.org)](https://jupyter.org) is the
core platform we will be using in this course and
is a popular platform for data scientists around the world.  We have a JupyterLab
setup for this course so that we can operate in a cloud-hosted environment, free from
some of the resource constraints of running Jupyter on your local machine (though you are free to set
it up on your own and seek my advice if you desire).

You have been given the information about the  Jupyter environment we have setup for our course, and
the underlying Python environment will be using is the [Anaconda (https://anaconda.com)](https://anaconda.com)
distribution.  It is not necessary for this assignment, but you are free to look at the multitude
of packages installed with Anaconda, though we will not use the majority of them explicitly.

As you will soon find out, Notebooks are an incredibly effective way to mix code with narrative
and you can create cells that are entirely code or entirely Markdown.  Markdown (MD or `md`) is
a highly readable text format that allows for easy documentation of text files, while allowing
for HTML-based rendering of the text in a way that is style-independent.

We will be using Markdown frequently in this course, and you will learn that there are many different
"flavors" or Markdown.  We will only be using the basic flavor, but you will benefit from exploring
the "Github flavored" Markdown, though you will not be responsible for using it in this course -- only the
"basic" flavor.  Please refer to the original course announcement about Markdown.

&#167;  **THERE IS NOTHING TO TURN IN FOR THIS PART.** Play with and become familiar with the basic functions of
the Lab environment given to you online in the course Blackboard.


&#167;  **PLEASE _CREATE A MARKDOWN DOCUMENT_ CALLED `semester_goals.md` WITH 3 SENTENCES/FRAGMENTS THAT
ANSWER THE FOLLOWING QUESTION:**

* **What do you wish to accomplish this semester in Data Mining?**

Read the documentation for basic Markdown [here](https://www.markdownguide.org/basic-syntax). 
Turn in the text `.md` file *not* the processed `.html`.  In whatever you turn in, 
you must show the use of *ALL* the following:

* headings (one level is fine),
* bullets,
* bold and italics

Again, the content of your document needs to address the question above and it should live
in the top level directory of your assignment submission.  This part will be graded but no
points are awarded for your answer.



### (0%) Familiarize yourself with Github and basic git 

[Github (https://github.com)](https://github.com) is the _de facto_ platform for open source software in the world based
on the very popular [git (https://git-scm.org)](https://git-scm.org) version control system. Git has a sophisticated set
of tools for version control based on the concept of local repositories for fast commits and remote
repositories only when collaboration and remote synchronization is necessary.  Github enhances git by providing
tools and online hosting of public and private repositories to encourage and promote sharing and collaboration.
Github hosts some of the world's most widely used open source software.

**If you are already familiar with git and Github, then this part will be very easy!**

&#167;  **CREATE A PUBLIC GITHUB REPO NAMED `"mcis6273-F21-datamining"` AND PLACE A README.MD FILE IN IT.**
Create your first file called
`README.md` at the top level of the repository.  You can put whatever text you like in the file 
(If you like, use something like [lorem ipsum](https://lipsum.com/)
to generate random sentences to place in the file.).
Please include the link to **your** Github repository that now includes the minimal `README.md`. 
You don't have to have anything elaborate in that file or the repo. 



### (0%) Explore JupyterHub Linux console integrating what you learned in the prior parts of this homework 

The Linux console in JupyterLab is a great way to perform command-line tasks and is an essential tool
for basic scripting that is part of a data scientist's toolkit.  Open a console in the lab environment
and familiarize yourself with your files and basic commands using git as indicated below.

1. In a new JupyterLab command line console, run the `git clone` command to clone the new
  repository you created in the prior part.
  You will want to read the documentation on this 
  command (try here [https://www.git-scm.com/docs/git-clone](https://www.git-scm.com/docs/git-clone) to get a good
  start).
2. Within the same console, modify your `README.md` file, check it in and push it back to your repository, using
  `git push`.  Read the [documentation about `git push`](https://git-scm.com/docs/git-push).
3. The commands `wget` and `curl` are useful for grabbing data and files from remote resources off the web.
  Read the documentation on each of these commands by typing `man wget` or `man curl` in the terminal.
  Make sure you pipe the output to a file or use the proper flags to do so.

&#167;  **THERE IS NOTHING TO TURN IN FOR THIS PART.**



### (30%) Listen to the Talk Python['Podcast'] from June 25, 2021: A Path to Data Science Interview with Sanyam Bhutani 

Data science is one of the most important and "hot" disciplines today
and there is a lot going on from data engineering to modeling and
analysis.

Bhutani is one of the top [Kaggle]() leaders and in this interview
shares his experience from computer science to data science, 
documenting some of the lessons he learned along the way.

Please listen to this one hour podcast and answer some of the questions below.
You can listen to it from one of the two links below:

* [Talk Python['Podcast'] landing page](https://talkpython.fm/episodes/transcript/322/a-path-into-data-science)
* [direct link to mp3 file](https://downloads.talkpython.fm/podcasts/talkpython/322-starting-in-data-sci.mp3)

&#167;  **PLEASE ANSWER THE FOLLOWING QUESTIONS AFTER LISTENING TO THE PODCAST:**

1. List 3 things that you learned from this podcast?

2. What is your reaction to the podcast?  Pick at least one point Sanyam brought up in the interview that you agree with and list your reason why.

3. After listening to the podcast, do you think you are more interested or less interested in a career in Data Science?


# 1. List 3 things that you learned from this podcast?

* Sanyam Bhutani use Sentry on all the talk python and web properties.They actually fixed a bug triggered by a user and had the upgrade ready to roll out and he can see the error and have already rolled out to fix.
* Create your Sentry account at 'Talk python.FM/Sentry'.And if you sign up with a code talk python 2021.Its good for 2months of entries team plan,which will give you upto 20 times as many monthly events as well as other features.
* Another one Sanyam Bhutani wrote a blog post of Kaggle competition in experince writing and he use to set goals for every year.

# 2. What is your reaction to the podcast?  Pick at least one point Sanyam brought up in the interview that you agree with and list your reason why.

* It was a very nice interview Sanyam Bhutani gave about data science and as he said kaggle is better to get started and build knowledge beyond it.

# 3. After listening to the podcast, do you think you are more interested or less interested in a career in Data Science?

* Yes I am more intrested in taking data science course as Sanyam Bhutani said we get many oppurtunities in data science or to get into python and programming

### (70%) Explore Python for data munging and analysis, with an introduction to CSV and Pandas 


Python's strengths shine when tasked with data munging and analysis.  As we will learn throughout
the course, there are a number of excellent data sources for open data of all kinds now
available for the public.  These open data sources are heralding the new era of transparency
from all levels from small municipal data to big government data, from transportation, to science,
to education.

To warm up to such datasets, we will be working with an interesting
dataset from the US Fish and Wildlife Service (FWS).  This is a 
water quality data set taken from a managed national refuge in
Virginia called Back Bay National Wildlife Refuge, which was 
established in 1938.  As a function of being managed by the FWS, 
water quality samples are taken regularly from the marshes within 
the refuge.

You can (and should) learn a little more about Back Bay from
this link, since it has an interesting history, features and wildlife.

* [https://www.fws.gov/refuge/Back_Bay/about.html](https://www.fws.gov/refuge/Back_Bay/about.html)


The data we will be looking at can be found as a direct download 
from data.gov, the US data repository where many datasets from
a variety of sources can be found -- mostly related to the 
multitude of US government agencies.

The dataset is a small water quality dataset with several decades
of water quality data from Back Bay.  We will be warming up
to this dataset with a basic investigation into the shape, content
and context of the data contained therein.


In this part of the assignment, we will make use of Python libraries to pull the data from the
endpoint and use [Pandas](https://pandas.pydata.org) to plot the data.  The raw CSV data is
readily imported into Pandas from the following URL:

* [FWS Water Quality Data 12/20/2020](https://catalog.data.gov/dataset/water-quality-data/resource/f4d736fd-ade9-4e3f-b8e0-ae7fd98b2f87)

Please take a look at the page, on it you will notice a link 
to the raw CSV file:

* [https://ecos.fws.gov/ServCat/DownloadFile/173741?Reference=117348](https://ecos.fws.gov/ServCat/DownloadFile/173741?Reference=117348)

We are going to explore this dataset to learn a bit more about the 
water quality characteristics of Bay Bay over the past couple decades
or so.

&#167;  **WRITE THE CODE IN YOUR NOTEBOOK TO LOAD AND RESHAPE THE COMPLETE CSV WATER QUALITY DATASET**:

You will need to perform the following steps:

1. **use [`pandas.read_csv()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) method to load the dataset** into a Pandas DataFrame;
2. **clean the data so that the range of years is restricted to the 20 year period from 1999 to 2018**
5. **store the entire dataset back into a new CSV** file called `back_bay_1998-2018_clean.csv`.

**HINTS:** _Here are some a code hints you might like to study and use to craft a solution:_

* study  [`pandas.DataFrame.query()]`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html?highlight=query#pandas.DataFrame.query) to learn how to filter and query year ranges
* study  [`pandas.DataFrame.groupby()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html?highlight=groupby#pandas.DataFrame.groupby) to understand how to group data


In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('BKB_WaterQualityData_2020084.csv')
df.head() #head of the data sheet

Site_Id Unit_Id  Read_Date  Salinity (ppt)  Dissolved Oxygen (mg/L)  \
0     Bay     NaN   1/3/1994             1.3                     11.7   
1     Bay     NaN  1/31/1994             1.5                     12.0   
2     Bay     NaN   2/7/1994             1.0                     10.5   
3     Bay     NaN  2/23/1994             1.0                     10.1   
4     Bay     NaN  2/28/1994             1.0                     12.6   

   pH (standard units)  Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  \
0                  7.3              0.40             0.40              5.9   
1                  7.4              0.20             0.35              3.0   
2                  7.2              0.25             0.60              5.9   
3                  7.4              0.35             0.50             10.0   
4                  7.2              0.20             0.40              1.6   

   Air Temp-Celsius  Air Temp (?F) Time (24:00) Field_Tech DateVerified  \
0               8.0          46.40        11:00        NaN          NaN   
1               2.6          36.68        11:30        NaN          NaN   
2               7.6          45.68         9:45        NaN          NaN   
3               2.7          36.86          NaN        NaN          NaN   
4               0.0          32.00        10:30        NaN          NaN   

  WhoVerified  AirTemp (C)  Year  
0         NaN          8.0  1994  
1         NaN          2.6  1994  
2         NaN          7.6  1994  
3         NaN          2.7  1994  
4         NaN          0.0  1994

In [3]:
df.tail() #Tail of the dataset

Site_Id Unit_Id   Read_Date  Salinity (ppt)  Dissolved Oxygen (mg/L)  \
2366     Bay     NaN  10/11/2018             1.9                      5.0   
2367     Bay     NaN  10/24/2018             0.0                      9.0   
2368     Bay     NaN  10/28/2018             0.9                      2.9   
2369     Bay     NaN   11/7/2018             1.7                      NaN   
2370     Bay     NaN  12/11/2018             0.1                      NaN   

      pH (standard units)  Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  \
2366                  7.0              4.00              1.2             25.0   
2367                  7.0              0.30              0.6             18.0   
2368                  7.0              0.40              0.9             13.0   
2369                  7.0              0.45              0.9             20.0   
2370                  7.0              0.10              0.1             10.0   

      Air Temp-Celsius  Air Temp (?F) Time (24:00) Field_Tech DateVerified  \
2366               NaN           78.0        09:30    Sue Poe   11/13/2019   
2367               NaN           58.0        09:30    Sue Poe   11/13/2019   
2368               NaN           49.0        09:20    Sue Poe   11/13/2019   
2369               NaN           65.0        09:45    Sue Poe   11/13/2019   
2370               NaN           42.0        09:40    Sue Poe   11/13/2019   

          WhoVerified  AirTemp (C)  Year  
2366  Christine Folks    25.555556  2018  
2367  Christine Folks    14.444444  2018  
2368  Christine Folks     9.444444  2018  
2369  Christine Folks    18.333333  2018  
2370  Christine Folks     5.555556  2018

In [4]:
df1 = df.query('Year >= 1999 & Year <= 2018')

In [5]:
df1.head()

Site_Id Unit_Id   Read_Date  Salinity (ppt)  Dissolved Oxygen (mg/L)  \
345       A     NaN   8/11/1999             3.0                     2.60   
346       A     NaN   8/18/1999             3.0                     1.60   
347       A     NaN    9/8/1999             1.8                     1.35   
348       A     NaN  10/15/1999             0.1                     5.65   
349       A     NaN  11/10/1999             0.0                     6.70   

     pH (standard units)  Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  \
345                  7.7              0.60             0.60             26.0   
346                  8.7               NaN              NaN             31.0   
347                  8.2              0.40             1.00             28.0   
348                  7.3              0.60             0.80             20.0   
349                  7.6              0.65             0.95             18.0   

     Air Temp-Celsius  Air Temp (?F) Time (24:00)    Field_Tech DateVerified  \
345               NaN          80.96        23:02  Not Recorded          NaN   
346               NaN          91.40         9:36  Not Recorded          NaN   
347               NaN          76.60        14:24  Not Recorded          NaN   
348               NaN            NaN         0:00  Not Recorded          NaN   
349               NaN            NaN         0:00  Not Recorded          NaN   

    WhoVerified  AirTemp (C)  Year  
345         NaN    27.200000  1999  
346         NaN    33.000000  1999  
347         NaN    24.777778  1999  
348         NaN   -17.777778  1999  
349         NaN   -17.777778  1999

In [6]:
df1.tail()

Site_Id Unit_Id   Read_Date  Salinity (ppt)  Dissolved Oxygen (mg/L)  \
2366     Bay     NaN  10/11/2018             1.9                      5.0   
2367     Bay     NaN  10/24/2018             0.0                      9.0   
2368     Bay     NaN  10/28/2018             0.9                      2.9   
2369     Bay     NaN   11/7/2018             1.7                      NaN   
2370     Bay     NaN  12/11/2018             0.1                      NaN   

      pH (standard units)  Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  \
2366                  7.0              4.00              1.2             25.0   
2367                  7.0              0.30              0.6             18.0   
2368                  7.0              0.40              0.9             13.0   
2369                  7.0              0.45              0.9             20.0   
2370                  7.0              0.10              0.1             10.0   

      Air Temp-Celsius  Air Temp (?F) Time (24:00) Field_Tech DateVerified  \
2366               NaN           78.0        09:30    Sue Poe   11/13/2019   
2367               NaN           58.0        09:30    Sue Poe   11/13/2019   
2368               NaN           49.0        09:20    Sue Poe   11/13/2019   
2369               NaN           65.0        09:45    Sue Poe   11/13/2019   
2370               NaN           42.0        09:40    Sue Poe   11/13/2019   

          WhoVerified  AirTemp (C)  Year  
2366  Christine Folks    25.555556  2018  
2367  Christine Folks    14.444444  2018  
2368  Christine Folks     9.444444  2018  
2369  Christine Folks    18.333333  2018  
2370  Christine Folks     5.555556  2018

In [7]:
df1.to_csv(r'back_bay_1998-2018_clean.csv', index = False, header=True)

&#167;  **USE PANDAS TO LOAD THE CSV DATA TO A DATAFRAME AND ANSWER THE FOLLOWING QUESTIONS:**

1. How many and what are the names of the columns in this dataset?
2. What is the mean `Dissolved Oxygen (mg/L)` over the entire dataset?
3. Which year were the highest number of `AirTemp (C)` data points collected?
4. Which year were the least number of `AirTemp (C)` data points collected?


To answer these questions, you'll need to dive further into Pandas, which is
the standard tool in the Python data science stack for loading, manipulating,
transforming, analyzing and preparing data as input to other tools such as
[Numpy (http://www.numpy.org/)](http://www.numpy.org/), 
[SciKitLearn (http://scikit-learn.org/stable/index.html)](http://scikit-learn.org/stable/index.html), 
[NLTK (http://www.nltk.org/)](http://www.nltk.org/) and others.

For this assignment, you will only need to learn how to load and select data using Pandas.

* **LOADING DATA**
The core data structure in Pandas is the `DataFrame`.  You will need to visit
the Pandas documentation [(https://pandas.pydata.org/pandas-docs/stable/reference/)](https://pandas.pydata.org/pandas-docs/stable/reference/)
to learn more about the library, but to help you along with a hint, read the
documentation on the [`pandas.read_csv()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) method.

* **SELECTING DATA**
The [tutorial here on indexing and selecting](http://pandas.pydata.org/pandas-docs/stable/indexing.html)
should be of great use in understanding how to index and select subsets of
the data to answer the questions.

* **GROUPING DATA** You may use [`DataFrame.value_counts()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.value_counts.html?highlight=value_counts#pandas.DataFrame.value_counts) or [`DataFrame.groupby()`](https://pandas.pydata.org/pandas-docs/stable/reference/groupby.html) to group
the data you need for these questons.  You will also find [`DataFrame.groupby()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html?highlight=groupby#pandas.DataFrame.groupby) and [`DataFrame.describe()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html?highlight=describe#pandas.DataFrame.describe) very useful.

**CODE HINTS**

Here is example code that should give you clues about the structure
of your code for this part.

```python
  import pandas as pd

  df = pd.read_csv('your_json_file.csv')

  # code for question 1 ... and so on
```
 


In [8]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1956 entries, 345 to 2370
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Site_Id                  1956 non-null   object 
 1   Unit_Id                  32 non-null     object 
 2   Read_Date                1956 non-null   object 
 3   Salinity (ppt)           1867 non-null   float64
 4   Dissolved Oxygen (mg/L)  1150 non-null   float64
 5   pH (standard units)      1872 non-null   float64
 6   Secchi Depth (m)         1891 non-null   float64
 7   Water Depth (m)          1890 non-null   float64
 8   Water Temp (?C)          1842 non-null   float64
 9   Air Temp-Celsius         46 non-null     float64
 10  Air Temp (?F)            1904 non-null   float64
 11  Time (24:00)             1905 non-null   object 
 12  Field_Tech               1956 non-null   object 
 13  DateVerified             386 non-null    object 
 14  WhoVerified           

In [9]:
df1.describe()

Salinity (ppt)  Dissolved Oxygen (mg/L)  pH (standard units)  \
count     1867.000000              1150.000000          1872.000000   
mean         0.410204                 5.867704             7.106490   
std          0.873802                 2.054290             0.799353   
min          0.000000                 0.000000             0.300000   
25%          0.000000                 4.300000             6.500000   
50%          0.000000                 5.900000             7.000000   
75%          0.100000                 7.300000             7.500000   
max          8.000000                11.200000             9.900000   

       Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  Air Temp-Celsius  \
count       1891.000000      1890.000000      1842.000000         46.000000   
mean           0.558958         0.768701        18.096417         14.682609   
std            0.467386         0.615332         8.413927         13.818784   
min            0.000000         0.010000         0.000000          3.000000   
25%            0.300000         0.400000        11.000000          9.000000   
50%            0.500000         0.645000        19.000000         11.000000   
75%            0.700000         1.000000        25.000000         16.000000   
max            7.400000         8.500000        74.000000         74.000000   

       Air Temp (?F)  AirTemp (C)         Year  
count    1904.000000  1956.000000  1956.000000  
mean       61.604748    15.537219  2008.134969  
std        15.668014    10.201950     5.419882  
min        10.500000   -17.777778  1999.000000  
25%        48.000000     8.888889  2004.000000  
50%        63.000000    16.666667  2008.000000  
75%        75.000000    23.888889  2013.000000  
max        91.400000    33.000000  2018.000000

Mean of Dissolved Oxygen (mg/L) is 5.867704 from df1.describe

In [10]:
df1.columns = [c.replace(' ', '_') for c in df1.columns]

In [11]:
df1.columns = [c.replace('(', '') for c in df1.columns]

In [12]:
df1.columns = [c.replace(')', '') for c in df1.columns]

In [13]:
df1.columns = [c.replace('/', '') for c in df1.columns]

In [14]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1956 entries, 345 to 2370
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Site_Id               1956 non-null   object 
 1   Unit_Id               32 non-null     object 
 2   Read_Date             1956 non-null   object 
 3   Salinity_ppt          1867 non-null   float64
 4   Dissolved_Oxygen_mgL  1150 non-null   float64
 5   pH_standard_units     1872 non-null   float64
 6   Secchi_Depth_m        1891 non-null   float64
 7   Water_Depth_m         1890 non-null   float64
 8   Water_Temp_?C         1842 non-null   float64
 9   Air_Temp-Celsius      46 non-null     float64
 10  Air_Temp_?F           1904 non-null   float64
 11  Time_24:00            1905 non-null   object 
 12  Field_Tech            1956 non-null   object 
 13  DateVerified          386 non-null    object 
 14  WhoVerified           386 non-null    object 
 15  AirTemp_C          

In [15]:
df1.query('AirTemp_C == 33')

Site_Id Unit_Id  Read_Date  Salinity_ppt  Dissolved_Oxygen_mgL  \
346       A     NaN  8/18/1999           3.0                   1.6   

     pH_standard_units  Secchi_Depth_m  Water_Depth_m  Water_Temp_?C  \
346                8.7             NaN            NaN           31.0   

     Air_Temp-Celsius  Air_Temp_?F Time_24:00    Field_Tech DateVerified  \
346               NaN         91.4       9:36  Not Recorded          NaN   

    WhoVerified  AirTemp_C  Year  
346         NaN       33.0  1999

1999 is the year highest number of AirTemp (C) data points collected

In [16]:
df1.query('AirTemp_C <= -17')

Site_Id Unit_Id   Read_Date  Salinity_ppt  Dissolved_Oxygen_mgL  \
348        A     NaN  10/15/1999           0.1                  5.65   
349        A     NaN  11/10/1999           0.0                  6.70   
354        B     NaN    9/8/1999           0.9                  3.60   
355        B     NaN  10/15/1999           1.0                  4.70   
357        B     NaN  11/10/1999           1.0                  6.30   
362        C     NaN    9/8/1999           1.0                  2.40   
364        C     NaN  10/15/1999           0.1                  5.50   
366        C     NaN  11/10/1999           0.2                 10.00   
371        D     NaN    9/8/1999           0.0                  1.50   
372        D     NaN   9/22/1999           0.0                  1.90   
373        D     NaN  10/15/1999           0.0                  3.25   
380      Bay     NaN   1/23/1999           1.1                 11.20   
381      Bay     NaN    8/5/1999           1.8                  1.80   
386      Bay     NaN  10/27/1999           0.3                  4.30   
393        A     NaN    4/3/2000           1.0                   NaN   
415        B     NaN    4/3/2000           1.5                   NaN   
437        C     NaN    4/3/2000           1.0                   NaN   
459        D     NaN    4/3/2000           0.0                  7.40   
481      Bay     NaN    4/3/2000           1.5                  7.60   
558        C     NaN   7/20/2001           NaN                   NaN   
587        D     NaN  10/22/2001           0.0                   NaN   
616        A     NaN   2/11/2002           2.0                   NaN   
635        A     NaN  12/16/2002           0.0                   NaN   
638        B     NaN   2/11/2002           2.0                   NaN   
655        B     NaN  12/16/2002           0.0                   NaN   
658        C     NaN   2/11/2002           2.0                  8.40   
677        C     NaN  12/16/2002           0.0                  5.60   
680        D     NaN   2/11/2002           0.0                  8.30   
699        D     NaN  12/16/2002           0.0                  5.20   
702      Bay     NaN   2/11/2002           3.0                  5.70   
865        B     NaN  11/18/2004           0.0                  5.30   
948        A     NaN    7/1/2005           0.0                   NaN   
949        A     NaN   7/14/2005           0.0                  2.20   
968        B     NaN    7/1/2005           0.0                   NaN   
969        B     NaN   7/14/2005           0.0                  3.50   
989        C     NaN    7/1/2005           0.0                  5.10   
990        C     NaN   7/14/2005           0.0                   NaN   
996        C     NaN  10/27/2005           0.0                  5.00   
1010       D     NaN    7/1/2005           0.0                  2.70   
1011       D     NaN   7/14/2005           0.0                   NaN   
1026     Bay     NaN    3/3/2005           1.0                   NaN   
1033     Bay     NaN    7/1/2005           0.0                  4.30   
1129       D     NaN   7/18/2006           0.0                  2.70   
1140     Bay     NaN    1/3/2006           0.0                  0.80   
1288       A     NaN   6/19/2008           1.0                  5.20   
1312       B     NaN   6/19/2008           1.0                  5.90   
1335       C     NaN   6/19/2008           2.0                   NaN   
1359       D     NaN   6/19/2008           1.0                   NaN   
1538       B     NaN    2/5/2010           NaN                   NaN   
1584       D     NaN   3/16/2010           NaN                   NaN   
1619     Bay     NaN  10/14/2010           0.0                  3.90   
2200       A     NaN  11/29/2017           NaN                   NaN   

      pH_standard_units  Secchi_Depth_m  Water_Depth_m  Water_Temp_?C  \
348                 7.3            0.60           0.80           20.0   
349                 7.6            0.65   

1999
2000
2001
2002
2004
2005
2006
2008
2010
2017
Above are the years least number of AirTemp (C) data points collected which is -17.777778

&#167;  **EXPLORING WATER SALINITY IN THE DATA**

The Back Bay refuge is on the eastern coast of Virginia and to
the east is the Atlantic Ocean.  Salinity is a measure of
the salt concentration of water, and you can learn a little more
about salinity in water [here](https://www.usgs.gov/special-topic/water-science-school/science/saline-water-and-salinity?qt-science_center_objects=0#qt-science_center_objects).

You will notice that there is a `Site_Id` variable in the data, which 
we will find refers to the five sampling locations (see the [documentation here](https://ecos.fws.gov/ServCat/Reference/Profile/117348))
of (1) the Bay, (2) D-Pool (fishing pond), (3) C-Pool, (4) B-Pool and (5) A-Pool.

The ppt in Salinity is the percent salinity, and so 1 ppt is equivalent to 10000 ppm salinity.  Use this information to answer 
the following questions.

1. Which sampling location has the highest mean ppt?  What is the equivalent ppm?
2. When looking at the mean ppt, which location would you infer is furthest from the influence of ocean water inflows? 
   (Assume that higher salinity correlates to closer proximity to the ocean.)
3. Dig a little deeper into #2, and write why there may be some uncertainty in your answer? (hint: certainty is improved by consistency in data)
4. Use the data to determine the correlation between `Salinity (ppt)` and `pH (standard units)`.  Use the [DataFrame.corr()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.corr.html?highlight=correlate).  You just need to report the correlation value. 

In [17]:
df1.query('Salinity_ppt == 0.4')

Site_Id Unit_Id   Read_Date  Salinity_ppt  Dissolved_Oxygen_mgL  \
374       D     NaN  10/27/1999           0.4                   3.4   

     pH_standard_units  Secchi_Depth_m  Water_Depth_m  Water_Temp_?C  \
374                7.4            0.95           1.35           42.5   

     Air_Temp-Celsius  Air_Temp_?F Time_24:00    Field_Tech DateVerified  \
374               NaN         72.0       0:00  Not Recorded          NaN   

    WhoVerified  AirTemp_C  Year  
374         NaN  22.222222  1999

Based on the df1.describe() mean of ppt is 0.4. Location D has the highest mean ppt and equivalent ppm is 3.4

In [18]:
df1.query('Salinity_ppt == 0')

Site_Id Unit_Id   Read_Date  Salinity_ppt  Dissolved_Oxygen_mgL  \
349        A     NaN  11/10/1999           0.0                  6.70   
365        C     NaN  10/27/1999           0.0                  3.70   
371        D     NaN    9/8/1999           0.0                  1.50   
372        D     NaN   9/22/1999           0.0                  1.90   
373        D     NaN  10/15/1999           0.0                  3.25   
...      ...     ...         ...           ...                   ...   
2362       D     NaN  10/24/2018           0.0                   NaN   
2363       D     NaN  10/28/2018           0.0                   NaN   
2364       D     NaN   11/7/2018           0.0                  6.90   
2365       D     NaN  12/11/2018           0.0                   NaN   
2367     Bay     NaN  10/24/2018           0.0                  9.00   

      pH_standard_units  Secchi_Depth_m  Water_Depth_m  Water_Temp_?C  \
349                 7.6            0.65           0.95           18.0   
365                 7.7            0.55           0.55           43.0   
371                 7.4            1.75           1.85           34.5   
372                 7.6            0.50           1.65           18.0   
373                 7.3            0.55           1.55           17.0   
...                 ...             ...            ...            ...   
2362                6.5            0.60           1.30           16.0   
2363                6.5            1.10           1.20           13.0   
2364                6.5            0.90           1.30           20.0   
2365                6.5            1.20           1.40           11.0   
2367                7.0            0.30           0.60           18.0   

      Air_Temp-Celsius  Air_Temp_?F Time_24:00    Field_Tech DateVerified  \
349                NaN          NaN       0:00  Not Recorded          NaN   
365                NaN         72.0       0:00  Not Recorded          NaN   
371                NaN          NaN      12:00  Not Recorded          NaN   
372                NaN          NaN       0:00  Not Recorded          NaN   
373                NaN          NaN       0:00  Not Recorded          NaN   
...                ...          ...        ...           ...          ...   
2362               NaN         58.0      10:15       Sue Poe   11/12/2019   
2363               NaN         49.0      10:15       Sue Poe   11/12/2019   
2364               NaN         65.0      10:22       Sue Poe   11/12/2019   
2365               NaN         42.0      10:15       Sue Poe   11/12/2019   
2367               NaN         58.0      09:30       Sue Poe   11/13/2019   

          WhoVerified  AirTemp_C  Year  
349               NaN -17.777778  1999  
365               NaN  22.222222  1999  
371               NaN -17.777778  1999  
372               NaN -17.777778  1999  
373               NaN -17.777778  1999  
...               ...        ...   ...  
2362  Christine Folks  14.444444  2018  
2363  Christine Folks   9.444444  2018  
2364  Christine Folks  18.333333  2018  
2365  Christine Folks   5.555556  2018  
2367  Christine Folks  14.444444  2018  

[1391 rows x 17 columns]

Many locations has Salinity ppt with 0 value Site id's are A,C,D & Bay

In [19]:
df1.query('Salinity_ppt == 0 & Dissolved_Oxygen_mgL == 0')

Site_Id Unit_Id  Read_Date  Salinity_ppt  Dissolved_Oxygen_mgL  \
1937       A   01CSV  6/10/2014           0.0                   0.0   
1938       B   01CSV  6/10/2014           0.0                   0.0   

      pH_standard_units  Secchi_Depth_m  Water_Depth_m  Water_Temp_?C  \
1937                7.0             0.5            0.7           27.0   
1938                8.0             0.3            0.3           27.0   

      Air_Temp-Celsius  Air_Temp_?F Time_24:00 Field_Tech DateVerified  \
1937              23.0         74.0      11:00    Sue Poe    6/23/2014   
1938              74.0         23.0      10:50    Sue Poe    6/23/2014   

         WhoVerified  AirTemp_C  Year  
1937  Trenton Miller  23.333333  2014  
1938  Trenton Miller  -5.000000  2014

Solubility of dissolved oxygen also decreases as salinity increases, based on this the dissolved oxygen with minimum is 0 and salinity 0 has Site ID with A & B

In [20]:
df1.corr()

Salinity_ppt  Dissolved_Oxygen_mgL  pH_standard_units  \
Salinity_ppt              1.000000             -0.096862           0.271262   
Dissolved_Oxygen_mgL     -0.096862              1.000000          -0.069449   
pH_standard_units         0.271262             -0.069449           1.000000   
Secchi_Depth_m           -0.077506              0.096643          -0.130557   
Water_Depth_m            -0.119635              0.054607          -0.182553   
Water_Temp_?C             0.008669             -0.564673           0.174527   
Air_Temp-Celsius               NaN             -0.512446           0.505689   
Air_Temp_?F               0.066299             -0.560187           0.191483   
AirTemp_C                 0.039016             -0.411543           0.116325   
Year                     -0.416750              0.108474          -0.360782   

                      Secchi_Depth_m  Water_Depth_m  Water_Temp_?C  \
Salinity_ppt               -0.077506      -0.119635       0.008669   
Dissolved_Oxygen_mgL        0.096643       0.054607      -0.564673   
pH_standard_units          -0.130557      -0.182553       0.174527   
Secchi_Depth_m              1.000000       0.810988      -0.070021   
Water_Depth_m               0.810988       1.000000      -0.008880   
Water_Temp_?C              -0.070021      -0.008880       1.000000   
Air_Temp-Celsius           -0.144087      -0.076694       0.734458   
Air_Temp_?F                -0.079466      -0.019103       0.860244   
AirTemp_C                  -0.084628      -0.027647       0.692660   
Year                        0.013742       0.038773      -0.058214   

                      Air_Temp-Celsius  Air_Temp_?F  AirTemp_C      Year  
Salinity_ppt                       NaN     0.066299   0.039016 -0.416750  
Dissolved_Oxygen_mgL         -0.512446    -0.560187  -0.411543  0.108474  
pH_standard_units             0.505689     0.191483   0.116325 -0.360782  
Secchi_Depth_m               -0.144087    -0.079466  -0.084628  0.013742  
Water_Depth_m                -0.076694    -0.019103  -0.027647  0.038773  
Water_Temp_?C                 0.734458     0.860244   0.692660 -0.058214  
Air_Temp-Celsius              1.000000    -0.204001  -0.204001 -0.646594  
Air_Temp_?F                  -0.204001     1.000000   1.000000 -0.065935  
AirTemp_C                    -0.204001     1.000000   1.000000  0.030111  
Year                         -0.646594    -0.065935   0.030111  1.000000

Correlation between Salinity & pH is 0.271262